### PR TITLE
Adding fuzz tests for shared directory eventual consistency issues,

### PR DESCRIPTION
**Pull request by jatgarg**
## Description

Adding fuzz tests for shared directory eventual consistency issues,

### CODE REVIEW

File: packages/dds/map/src/test/mocha/directoryFuzzTests.spec.ts [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+            `key: ${key}, value in dir ${first.id}: ${first.get(key)} and in ${second.id}: ${second.get(key)}`);
+    }
+
+    // Check for number of subdirectores with both directories.
+    assert.strictEqual(first.countSubDirectory(), second.countSubDirectory(),
+        `Number of subDirectories not same: Number of subdirectory in ` +
+        `${first.id}: ${first.countSubDirectory()} and in ${second.id}: ${second.countSubDirectory()}`);
+
+    // Check for consistency of subdirectores with both directories.
+    for (const [name, subDirectory1] of first.subdirectories()) {
```
Comment (Abe27342) [inline packages/dds/map/src/test/mocha/directoryFuzzTests.spec.ts:445]: i'm not particularly familiar with the subdirectory API, but it seems like you either want to do a length assertion on size of subdirectories matching OR validate that there aren't subdirectories in the second that are in the first here.

Comment (jatgarg) [reply packages/dds/map/src/test/mocha/directoryFuzzTests.spec.ts:445]: So, a subdirectory could have key, value pairs and sub directories. So, I am matching the size which is no. of keys in it and then check that all key value pairs match. And then I am checking number of sub directories and then recursively checking eventual consistency on those sub directories.

File: packages/dds/map/src/test/mocha/directoryFuzzTests.spec.ts [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+
+interface OperationGenerationConfig {
+    validateInterval: number,
+    maxSubDirectoryChild?: number;
+    subDirectoryNamePool?: string[];
+    keyNamePool?: string[];
+}
+
+const defaultOptions: Required<OperationGenerationConfig> = {
+    validateInterval: 2,
```
Comment (Abe27342) [inline packages/dds/map/src/test/mocha/directoryFuzzTests.spec.ts:95]: i see that you override this default in the test below, but still probably worth making it more than 2. That would synchronize after every op.

File: packages/dds/map/src/test/mocha/directoryFuzzTests.spec.ts [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+    subDirectoryNamePool: ["dir1", "dir2", "dir3"],
+    keyNamePool: ["prop1", "prop2", "prop3"],
+};
+
+function makeOperationGenerator(optionsParam?: OperationGenerationConfig): Generator<Operation, FuzzTestState> {
+    const options = { ...defaultOptions, ...(optionsParam ?? {})};
+    type ClientOpState = FuzzTestState & { sharedDirectory: SharedDirectory; };
+
+    // All subsequent helper functions are generators; note that they don't actually apply any operations.
+    function pickAbsolutePathForCreateDirectoryOp(state: ClientOpState): string {
```
Comment (Abe27342) [inline packages/dds/map/src/test/mocha/directoryFuzzTests.spec.ts:106]: it feels like you could probably simplify the `pick` methods here by writing a generic method to pick a subdirectory and rejection sample it based on the op (e.g. for subdir create, reject if the selected directory is already at max size).

File: packages/dds/map/src/test/mocha/directoryFuzzTests.spec.ts [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+            path: pickAbsolutePathForCreateDirectoryOp(state),
+        }
+    }
+
+    function deleteSubDirectory(state: ClientOpState): DeleteSubDirectory {
+        const { random, sharedDirectory } = state;
+        const path = pickAbsolutePathForDeleteDirectoryOp(state);
+        const parentDir = sharedDirectory.getWorkingDirectory(path);
+        assert(parentDir !== undefined, "parent dir should be defined");
+        assert(parentDir.countSubDirectory && parentDir.countSubDirectory() > 0, "Atleast 1 sudir should be there");
```
Comment (Abe27342) [inline packages/dds/map/src/test/mocha/directoryFuzzTests.spec.ts:208]:
```suggestion
        assert(parentDir.countSubDirectory && parentDir.countSubDirectory() > 0, "Atleast 1 subdir should be there");
```

**Review by Abe27342 — APPROVED** [review]
Overall this looks good. I have a couple nits, but excited to see the added coverage.

I also filed [this](https://dev.azure.com/fluidframework/internal/_workitems/edit/3211) for hooking these up to the dds stress pipeline.

File: packages/dds/map/src/test/mocha/directoryFuzzTests.spec.ts (hunk 2 above)
Comment (jatgarg) [reply packages/dds/map/src/test/mocha/directoryFuzzTests.spec.ts]: Changed it to 10 too.

File: packages/dds/map/src/test/mocha/directoryFuzzTests.spec.ts (hunk 3 above)
Comment (jatgarg) [reply packages/dds/map/src/test/mocha/directoryFuzzTests.spec.ts:106]: I combined the methods for generating key ops. There seems to be a good diff between the dir ops one, so keeping them separate for now.

**Pull request merged**